src/lib.rs

```diff
@@ -1,6 +1,138 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
+use std::io::Write;
+use std::net::SocketAddrV4;
+use std::str::FromStr;
+
+pub struct Finder{
+    name: String,
+    socket: Option<SocketAddrV4>,
+    beacon_receiving_socket: std::net::UdpSocket,
+    last_beacon_time: std::time::Instant
+}
+
+pub struct Beacon{
+    data: Vec<u8>
+}
+
+pub enum Status{
+    Found(SocketAddrV4),
+    TimeSince(std::time::Duration)
+}
+
+struct Target<'a>{
+    name: &'a str,
+    source_address: SocketAddrV4,
+    port: u16
+}
+
+pub fn new(name: String) -> Result<Finder, std::io::Error>{
+    let last_beacon_time = std::time::Instant::now();
+    let beacon_receiving_socket = std::net::UdpSocket::bind("0.0.0.0:9092")?;
+    beacon_receiving_socket.set_nonblocking(true);
+    Ok( Finder{name, socket: None, last_beacon_time, beacon_receiving_socket} )
+}
+
+pub fn beacon(name: &str, address: u16) -> Beacon{
+    let mut data = vec![0u8; name.len() + 8];
+    write!(&mut data[..], "-{}-{}-", name, address);
+    Beacon{data}
+}
+
+impl Finder{
+    fn check_socket(&self) -> Option<(std::net::SocketAddrV4, String)>{
+
+        let mut data_found = None;
+        let mut source_address: SocketAddrV4;
+        let mut buffer = [0u8; 500];
+
+        let mut latest_message = None;
+        loop{
+            match self.beacon_receiving_socket.recv_from(&mut buffer){
+                Ok(packet) => {
+                    latest_message = Some(packet);
+                }
+                Err(error) => {
+                    break;
+                }
+            }
+        }
+        if let Some(packet) = latest_message{
+            source_address = match packet.1{
+                std::net::SocketAddr::V4(src) => src,
+                _ => panic!("V6 not supported")
+            };
+            let data = String::from_utf8(buffer.to_vec()).unwrap();
+            data_found = Some( (source_address, data) );
+        }
+        data_found
+    }
+
+    fn extract_message<'a>(&self, data: &'a str) -> (&'a str, u16){
```

Comment:
This method doesn't use `self` anywhere so probs makes more sense as a free function. I think this would also remove the need for the explicit lifetime as there would be no longer be any ambiguity about which input reference the returned `&str` is bound to.